akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala

```diff
@@ -106,13 +109,30 @@ sealed trait HttpMessage extends jm.HttpMessage {
   /** Returns the first header of the given type if there is one */
   def header[T >: Null <: jm.HttpHeader: ClassTag]: Option[T] = {
     val clazz = classTag[T].runtimeClass.asInstanceOf[Class[T]]
-    HttpHeader.fastFind[T](clazz, headers) match {
+    if (classOf[akka.http.scaladsl.model.headers.ModeledCustomHeader[T]].isAssignableFrom(clazz)) {
+      findHeaderViaModeledCompanion(clazz).asInstanceOf[Option[T]]
+    } else HttpHeader.fastFind[T](clazz, headers) match {
       case OptionVal.Some(h)                     ⇒ Some(h)
       case _ if clazz == classOf[`Content-Type`] ⇒ Some(`Content-Type`(entity.contentType)).asInstanceOf[Option[T]]
       case _                                     ⇒ None
     }
   }
 
+  private def findHeaderViaModeledCompanion(clazz: Class[_]) = {
+    // allows locating "companion" without implicits, same trick is applied in javadsl
```

Comment:
I wonder if it wouldn't make more sense to try to find better ways in the long-term which use implicits for the same effect.